### DISCUSSION TITLE
create a location picker based on leaflet (in a repeater leaflet doesn't initiate properly)

### DISCUSSION THREAD

**Discussion by ehs4nm** [filamentphp/filament#2602] [Help]
Hi. I am trying to make custom field as a location picker (lat, lng) using [leaflet](https://leafletjs.com/reference.html) library.
Using it in a default resource-page form is OK but going further make some trouble.
First In a repeater eg, HasManyRepeater when state is null (model hasn't been created yet) it makes error.
In HasManyRepeater when there are more than one map all the maps can't be initiated. i used a random id for each one but couldn't succeed.
Second in relation-managers when the modal opens the leaflet can't initiate map again.

here is my code:
`
{{-- @php //commented out
  $random_map_id = rand(1,99); //generating random map id
@endphp --}}

@once  {{-- loading the leaflet Js/Css--}}
    @push('scripts')
        <link rel="stylesheet" href="https://unpkg.com/leaflet@1.8.0/dist/leaflet.css"
        integrity="sha512-hoalWLoI8r4UszCkZ5kL8vayOGVae1oxXe/2A4AO6J9+580uKHDO3JdHb7NzwwzK5xr/Fs0W40kiNHxM9vyTtQ=="
        crossorigin=""/>
        <script src="https://unpkg.com/leaflet@1.8.0/dist/leaflet.js"
        integrity="sha512-BB3hKbKWOc9Ez/TAwyWxNXeoV9c1v6FIeYiBieIWkpLjauysF18NzgR1MBNBXf8/KABdlkX68nAhlwcDFLGPCQ=="
        crossorigin=""></script>
        <script>
            var map = L.map('map').setView([35.7219, 51.3347], 13);
            var tiles = L.tileLayer('https://{s}.tile.openstreetmap.org/{z}/{x}/{y}.png', {
                maxZoom: 19,
                attribution: '&copy; <a href="http://www.openstreetmap.org/copyright">OpenStreetMap</a>'
            }).addTo(map);         
        </script>
    @endpush
@endonce

<x-forms::field-wrapper
    :id="$getId()"
    :label="$getLabel()"
    :label-sr-only="$isLabelHidden()"
    :helper-text="$getHelperText()"
    :hint="$getHint()"
    :hint-icon="$getHintIcon()"
    :required="$isRequired()"
    :state-path="$getStatePath()"
    >

    <div wire:ignore x-data="{ state: $wire.entangle('{{ $getStatePath() }}') }" @notify.document="state=$event.detail">
        <template x-if="state != null">
            <div class="p-3">
                <span>lat: <small  x-text="state.lat"></small></span>
                <span>lng: <small  x-text="state.lng "></small></span>
            </div>
        </template>
    
        <div id="map" style="height: 300px; width:300px;"></div>
        <script>

        window.addEventListener('load', function () {
  
            
            marker = L.marker([35.7219, 51.3347], {draggable: true}).addTo(map);

            marker.on('dragend', function (event){
                var event = new CustomEvent("notify", { "detail": marker.getLatLng()});
                document.dispatchEvent(event);
            });
        })

        </script>

    </div>
</x-forms::field-wrapper>
`

**Comment by danharrin** [comment]
Since the `scripts` stack is in the layout is only rendered on the first page load as Livewire does not support dynamic stacks. So, if your `@push` if not there on the first page load, the library will not be loaded.

**Comment by ehs4nm** [comment]
So it must be loaded on all pages?
The date picker component included a JS lib (translation) the same way and it works.
where should i put this?

One thing more when i try to let a default value for setview of map like below

`        <script>
            var data = @this.data;
            if(data.latlng === null) data.latlng = {'lat': 35.7219, 'lng': 51.3347}; 
            var map = L.map('map').setView([data.latlng['lat'] , data.latlng['lng']], 13);
            var tiles = L.tileLayer('https://{s}.tile.openstreetmap.org/{z}/{x}/{y}.png', {
                maxZoom: 19,
                attribution: '&copy; <a href="http://www.openstreetmap.org/copyright">OpenStreetMap</a>'
            }).addTo(map);         
        </script>`

I will get: Livewire encountered corrupt data when trying to hydrate the [...] component. It seems the data state is changing!
